### DISCUSSION TITLE
Lazyvim automatically formats C functions to one line

### DISCUSSION THREAD

**Discussion by carnivorecookies** [LazyVim/LazyVim#4919] [Q&A] [answered]
My c file contains the following code:
```c
void stack_free(Stack *stack) {
  free(stack->arr);
}
```

When I save, lazyvim formats it to:

```c
void stack_free(Stack *stack) { free(stack->arr); }
```

What causes this, and how do I disable this formatting?

**Comment by ferid333** [comment]
Hello @darkk99 

This formatting is likely caused by a **code formatter** (e.g., `clang-format`) integrated with LazyVim. LazyVim runs formatters automatically on save, depending on its configuration and installed tools.

1. **Locate the Formatter Configuration in LazyVim**:
   - Open your LazyVim configuration (usually in `~/.config/nvim/lua` or a similar path).
   - Look for settings related to formatters. For LazyVim, it often uses `null-ls` for formatting.

2. **Disable Formatting for C Files**:
   - Add a rule to prevent formatting for C files. Example:
     ```lua
     local null_ls = require("null-ls")
     null_ls.setup({
       sources = {
         -- Add formatters here, but exclude C if clang-format is causing the issue
       },
       on_attach = function(client, bufnr)
         if vim.bo[bufnr].filetype == "c" then
           client.server_capabilities.documentFormattingProvider = false
         end
       end,
     })
     ```

3. **Disable Auto-Formatting Entirely**:
   - If you want to disable auto-formatting globally, you can set this in the LazyVim config:
     ```lua
     vim.api.nvim_create_autocmd("LspAttach", {
       callback = function(args)
         local client = vim.lsp.get_client_by_id(args.data.client_id)
         client.server_capabilities.documentFormattingProvider = false
       end,
     })
     ```

4. **Disable Specific Formatters**:
   - If `clang-format` is used, you can adjust its settings or exclude it from being called. Find where it's configured in your `LazyVim` setup.

---

**Reply by carnivorecookies** [reply]
i don't want to disable formatting altogether; just this specific part of the formatter

**Reply by ferid333** [reply, marked as answer]
Well, If you want to tweak the formatter to prevent it from reformatting specific parts of your code while keeping it enabled overall, here’s how you can approach it:

### 1. Adjust Your Code Formatter Configuration
If you’re using **`clang-format`** (commonly used for C code), you can modify its configuration file to prevent specific rules.

#### Example: `.clang-format` File
Add or update the `.clang-format` file in your project directory (or home directory). To prevent single-line formatting for function definitions:
```yaml
AllowShortFunctionsOnASingleLine: false
```

This setting forces `clang-format` to keep multi-line functions as-is.

---

### 2. Use Formatter-Specific Comments
`clang-format` supports disabling formatting for specific sections of code using **comments**:

```c
// clang-format off
void stack_free(Stack *stack) {
  free(stack->arr);
}
// clang-format on
```

This tells the formatter to skip the code block within the comments.

---

### 3. Fine-Tune LazyVim Formatter Settings
If `clang-format` is being invoked by LazyVim through tools like `null-ls` or `lsp`, you can customize how the formatter behaves:

- **Pass Custom Flags**:
  Modify the LazyVim formatter setup to include custom flags for `clang-format`:
  ```lua
  local null_ls = require("null-ls")
  null_ls.setup({
    sources = {
      null_ls.builtins.formatting.clang_format.with({
        extra_args = { "--style={AllowShortFunctionsOnASingleLine: false}" },
      }),
    },
  })
  ```

**Reply by carnivorecookies** [reply]
Putting
```yaml
AllowShortFunctionsOnASingleLine: false
```
in ~/.clang-format worked. Thank you!

However, `null-ls` is not used (by default) in LazyVim, so that didn't work.
